org.eclipse.dawnsci.analysis.dataset/src/org/eclipse/dawnsci/analysis/dataset/mask/MaskCircularBuffer.java

```diff
@@ -0,0 +1,180 @@
+package org.eclipse.dawnsci.analysis.dataset.mask;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.dawnsci.analysis.api.roi.IROI;
+import org.eclipse.dawnsci.analysis.api.roi.IRectangularROI;
+import org.eclipse.january.dataset.BooleanDataset;
+import org.eclipse.january.dataset.Comparisons;
+import org.eclipse.january.dataset.Dataset;
+import org.eclipse.january.dataset.DatasetFactory;
+import org.eclipse.january.dataset.IndexIterator;
+import org.eclipse.january.dataset.IntegerDataset;
+
+/**
+ * Buffer that allows multiple masking steps to be applied and undone
+ * 
+ * Optimised for speed and low memory usage
+ * 
+ */
+public class MaskCircularBuffer {
+
+	private IntegerDataset mask;
+	private int[] shape;
+	int bitMask = START;
+	private static final int END = 0x80000000;
+	private static final int START = 0x00000001;
+	
+	public MaskCircularBuffer(int[] shape) {
+		mask = DatasetFactory.zeros(IntegerDataset.class, shape);
+		this.shape = shape;
+	}
+	
+	public void maskROI(IROI roi) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 33 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)